fiftyone/brain/internal/core/leaky_splits.py

```diff
@@ -0,0 +1,421 @@
+        to_keep = []
+        for (
+            sample_id,
+            neighbors,
+        ) in self._similarity_index.neighbors_map.items():
+            keep_sample = False
+            sample_split = self._id2split.get(sample_id, None)
+            if sample_split is None:
+                _throw_index_is_bigger_warning(sample_id)
+                continue  # sample is in index but not passed to leaky_splits
+            leaks = []
+            for n in neighbors:
+                neighbor_split = self._id2split.get(n[0], None)
+                if neighbor_split is None:
+                    _throw_index_is_bigger_warning(n[0])
+                    continue
+                if not (neighbor_split == sample_split):
+                    # at least one of the neighbors is from a different split
+                    # we keep this one
+                    keep_sample = True
+                    leaks.append(n[0])
+
+            if keep_sample:
+                to_keep.append(sample_id)
+                # remove all other samples because they are all from the same split
+                to_keep = to_keep + leaks
+
+        duplicates = duplicates.select(to_keep)
+
+        # cache to avoid recomputation
+        self._last_computed_threshold = self._leak_threshold
+        self._leaks = duplicates
+
+        return duplicates
+
+    @property
+    def num_leaks(self):
+        """Returns the number of leaks found."""
+        return self.leaks.count
+
+    def leaks_for_sample(self, sample):
+        """Return view with all leaks related to a certain sample.
+
+        Args:
+            sample: sample object or sample id
+
+        """
+        # compute leaks if it hasn't happend yet
+        if not self._leak_threshold == self._last_computed_threshold:
+            _ = self.leaks
+
+        sample_id = sample if isinstance(sample, str) else sample["id"]
+        sample_split = self._id2split[sample_id]
+        neighbors_ids = []
+        if sample_id in self._similarity_index.neighbors_map.keys():
+            neighbors = self._similarity_index.neighbors_map[sample_id]
+            neighbors_ids = [
+                n[0]
+                for n in neighbors
+                if not self._id2split[n[0]] == sample_split
+            ]
+        else:
+            for (
+                unique_id,
+                neighbors,
+            ) in self._similarity_index.neighbors_map.items():
+                if sample_id in [n[0] for n in neighbors]:
+                    neighbors_ids = [
+                        n[0]
+                        for n in neighbors
+                        if not self._id2split[n[0]] == sample_split
+                    ]
+                    neighbors_ids.append(unique_id)
+                    break
+
+        return self.samples.select([sample_id] + neighbors_ids)
+
+    def no_leaks_view(self, view):
+        return view.exclude(self.leaks.values("id"))
+
+    def tag_leaks(self, tag="leak"):
```

Comment:
You will need to change `leaks` to `leaks_view` here as well. Also, instead of iterating over all samples and saving, please use [`tag_samples` ()](https://docs.voxel51.com/api/fiftyone.core.collections.html#fiftyone.core.collections.SampleCollection.tag_samples)